test/invariant/fuzz/helpers/Pools.t.sol

```diff
@@ -0,0 +1,91 @@
+
+import {DirectAllocationStrategy} from "contracts/strategies/examples/direct-allocation/DirectAllocation.sol";
+import {DonationVotingOnchain} from "contracts/strategies/examples/donation-voting/DonationVotingOnchain.sol";
+import {EasyRPGF} from "contracts/strategies/examples/easy-rpgf/EasyRPGF.sol";
+import {QVImpactStream} from "contracts/strategies/examples/impact-stream/QVImpactStream.sol";
+import {QVSimple} from "contracts/strategies/examples/quadratic-voting/QVSimple.sol";
+import {RFPSimple} from "contracts/strategies/examples/rfp/RFPSimple.sol";
+import {SQFSuperfluid} from "contracts/strategies/examples/sqf-superfluid/SQFSuperfluid.sol";
+
+contract Pools is Utils {
+    Allo private allo;
+
+    enum PoolStrategies {
+        None,
+        DirectAllocation,
+        DonationVoting,
+        EasyRPGF,
+        ImpactStream,
+        QuadraticVoting,
+        RFP,
+        SQFSuperfluid
+    }
+
+    uint256[] internal ghost_poolIds;
+    mapping(uint256 _poolId => address _poolAdmin) ghost_poolAdmins;
+
+    // mapping(uint256 _poolId => PoolStrategies _strategy) internal _poolStrategy;
+
+    mapping(PoolStrategies _strategy => address _implementation)
+        internal _strategyImplementations;
+
+    function _initImplementations(address _allo) internal {
+        _strategyImplementations[PoolStrategies.DirectAllocation] = address(
+            new DirectAllocationStrategy(_allo)
+        );
+        _strategyImplementations[PoolStrategies.DonationVoting] = address(
+            new DonationVotingOnchain(_allo, "MyFancyName")
+        );
+        _strategyImplementations[PoolStrategies.EasyRPGF] = address(
+            new EasyRPGF(_allo)
+        );
+        _strategyImplementations[PoolStrategies.ImpactStream] = address(
+            new QVImpactStream(_allo)
+        );
+        _strategyImplementations[PoolStrategies.QuadraticVoting] = address(
+            new QVSimple(_allo, "MyFancyName")
+        );
+        _strategyImplementations[PoolStrategies.RFP] = address(
+            new RFPSimple(_allo)
+        );
+        _strategyImplementations[PoolStrategies.SQFSuperfluid] = address(
+            new SQFSuperfluid(_allo)
+        );
+
+        allo = Allo(_allo);
+    }
+
+    function _recordPool(uint256 _poolId, PoolStrategies _strategy) internal {
+        // _poolStrategy[_poolId] = _strategy;
+        ghost_poolIds.push(_poolId);
+    }
+
+    // reverse lookup pool id -> strategy type
+    function _poolStrategy(uint256 _poolId) internal returns (PoolStrategies) {
+        IAllo.Pool memory _pool = allo.getPool(_poolId);
+        for (uint256 i; i < uint256(type(PoolStrategies).max); i++)
+            if (
+                _strategyImplementations[PoolStrategies(i)] ==
+                address(_pool.strategy)
+            ) return PoolStrategies(i);
+
+        emit TestFailure("Wrong pool strategy implementation address");
+    }
+
+    function _strategyHasImplementation(
```

Comment:
another option is to use an aux `mapping(strategy => implementation)` to avoid the for loop

Reply:
as we are not optimising for gas, I think we should avoid as much as possible having state in the test contract, when we can easily(ish) compute it instead - wdyt?

Reply:
yes it's fine for me